azure-iot-device/azure/iot/device/common/mqtt_transport.py

```diff
@@ -61,8 +115,18 @@ def _create_mqtt_client(self):
         def on_connect(client, userdata, flags, rc):
             logger.info("connected with result code: {}".format(rc))
 
-            # TODO: how to do failed connection?
-            if self.on_mqtt_connected:
+            if rc:
```

Comment:
>if rc: [](start = 12, length = 6)

I'm unclear on how (and why) this flow is different from the error handling based on the rc in the .`connect` method. Could you clarify?

Reply:
This has all been refactored, but a connection or reconnection can fail in two different ways:
1. If Paho fails before it can even send the CONNECT packet to the service, then the .connect method fails.
2. If Paho gets an error back from the server in the CONACK packet, then that error gets returned via the on_connected handler .

---
In reply to: [307882272](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r307882272) [](ancestors = 307882272)